### ISSUE TITLE
AnalyzerTest.RunAsync (Usable with CSharpSourceGeneratorTest and VisualBasicSourceGeneratorTest) should allow passing in expeced Diagnostics on tests.

### ISSUE THREAD

**Issue by AraHaan** [dotnet/roslyn-sdk#860]
In my Unit tests on my code I wanted to be able to pass that information to the RunAsync method in some way to tell it that I expect only specific diagnostics in that test, if not reported then fail the test but could not and I could only make it pass by using ``Assert.ThrowsAsync`` to avoid the failure exception on that test making it fail.

https://github.com/Elskom/GitBuildInfo.SourceGenerator/pull/37/files#diff-ab1f9e3ecaca5cd97cca73352bd17f37fe90532fbae37f11008c94b787e1dfa9R82-R94
https://github.com/Elskom/GitBuildInfo.SourceGenerator/pull/37/files#diff-ab1f9e3ecaca5cd97cca73352bd17f37fe90532fbae37f11008c94b787e1dfa9L151-R139
https://github.com/Elskom/GitBuildInfo.SourceGenerator/pull/37/files#diff-67aef68ff84e12d645ce41925d434bd146c1862bae2803413bebd0b7d5cad2faR13-R68

I think ``RunAsync`` needs an overload that can take an array (or list) of expected Diagnostic data and then verify that they were reported, otherwise fail.

**Comment by AraHaan** [comment]
This also explains another pain:
-when wanting to use the same code for both tests that expects diagnostics to be generated and tests that expects no diagnostics to be generated.

When doing that one could do:
```cs
private static async Task RunTest(string assemblyType, string isGeneric, string generatedSource, string testSource, DiagnosticResult expectedDiagnostic = default)
        {
            await new CSGeneratorTest
            {
                GlobalOptions =
                {
                    ("build_property.GitBuildInfoAssemblyType", assemblyType),
                    ("build_property.GitBuildInfoIsGeneric", isGeneric),
                    ("build_property.GitHead", "fbgtgretgtre"),
                    ("build_property.CommitHash", "vfdbttregter"),
                    ("build_property.GitBranch", "vsdfvfdsv"),
                },
                ReferenceAssemblies = ReferenceAssemblies.Net.Net50,
                TestState =
                {
                    Sources =
                    {
                        // do not duplicate the code to Elskom.GitInformation here.
                        File.ReadAllText(
                            "../../../../src/Elskom.GitInformation/GitInformationAttribute.cs"),
                        File.ReadAllText(
                            "../../../../src/Elskom.GitInformation/GitInformation.cs"),
                        testSource
                    },
                    GeneratedSources =
                    {
                        (typeof(SourceGenerator), "GitAssemblyInfo.g.cs", generatedSource)
                    },
                    ExpectedDiagnostics =
                    {
                        expectedDiagnostic
                    },
                },
            }.RunAsync();
        }
```
This would fail the tests expecting no diagnostic by making a Diagnostic saying that ```
Error Message:
   Microsoft.CodeAnalysis.Testing.Verifiers.EqualWithMessageException : Context: Diagnostics of test state
Mismatch between number of diagnostics returned, expected "1" actual "0"

Diagnostics:
    NONE.

Assert.Equal() Failure
Expected: 1
Actual:   0```

When the tests intended those to expect 0.

As such I think there should be another option then, which is to pass ``DiagnosticResult[]`` arrays into ``RunAsync`` for the expected results, while also working on the same codebase used in tests that expects no Diagnostics to be generated.

**Comment by jmarolf** [comment]
> When the tests intended those to expect 0.

I do not understand why you would expect the test to pass if you gave and expected diagnostic and none were found. Is this not an acceptable way to implement your helper method?

```csharp
private static async Task RunTest(
    string assemblyType,
    string isGeneric,
    string generatedSource,
    string testSource,
    DiagnosticResult? expectedDiagnostic = null)
{
    var test = new CSGeneratorTest
    {
        GlobalOptions =
        {
            ("build_property.GitBuildInfoAssemblyType", assemblyType),
            ("build_property.GitBuildInfoIsGeneric", isGeneric),
            ("build_property.GitHead", "fbgtgretgtre"),
            ("build_property.CommitHash", "vfdbttregter"),
            ("build_property.GitBranch", "vsdfvfdsv"),
        },
        ReferenceAssemblies = ReferenceAssemblies.Net.Net50,
        TestState =
        {
            Sources =
            {
                // do not duplicate the code to Elskom.GitInformation here.
                File.ReadAllText(
                    "../../../../src/Elskom.GitInformation/GitInformationAttribute.cs"),
                File.ReadAllText(
                    "../../../../src/Elskom.GitInformation/GitInformation.cs"),
                testSource
            },
            GeneratedSources =
            {
                (typeof(SourceGenerator), "GitAssemblyInfo.g.cs", generatedSource)
            },
        },
    };
    if (expectedDiagnostic is not null)
    {
        test.ExpectedDiagnostics.Add(expectedDiagnostic.Value);
    }

    await test.RunAsync();
}
```

**Comment by AraHaan** [comment]
That one works 🤔.
